algorithms/src/snark/marlin/ahp/prover/oracles.rs

```diff
@@ -14,45 +14,72 @@
+    /// Iterate over the polynomials output by the prover in the first round.
+    /// Intended for use when committing.
+    pub fn iter_for_commit(&'a self) -> impl Iterator<Item = LabeledPolynomialWithBasis<'a, F>> {
+        self.batches.iter().flat_map(|b| b.iter_for_commit()).chain(self.mask_poly.as_ref().map(Into::into))
+    }
+
+    /// Iterate over the polynomials output by the prover in the first round.
+    /// Intended for use when opening.
+    pub fn iter_for_open(&'a self) -> impl Iterator<Item = &'a LabeledPolynomial<F>> {
+        self.batches.iter().flat_map(|b| b.iter_for_open()).chain(self.mask_poly.as_ref())
+    }
+
+    pub fn matches_info(&self, info: &BTreeMap<PolynomialLabel, PolynomialInfo>) -> bool {
+        self.batches.iter().all(|b| b.matches_info(info))
+            && self.mask_poly.as_ref().map_or(true, |p| p.info() == &info[p.label()])
+    }
+}
+
+#[derive(Debug, Clone)]
+pub(in crate::snark::marlin) struct SingleEntry<'a, F: PrimeField> {
+    /// The evaluations of `Az`.
+    pub(super) z_a: LabeledPolynomialWithBasis<'a, F>,
+    /// The evaluations of `Bz`.
+    pub(super) z_b: LabeledPolynomialWithBasis<'a, F>,
     /// The LDE of `w`.
-    pub w_poly: LabeledPolynomial<F>,
+    pub(super) w_poly: LabeledPolynomial<F>,
     /// The LDE of `Az`.
-    pub z_a_poly: LabeledPolynomial<F>,
+    pub(super) z_a_poly: LabeledPolynomial<F>,
     /// The LDE of `Bz`.
-    pub z_b_poly: LabeledPolynomial<F>,
+    pub(super) z_b_poly: LabeledPolynomial<F>,
 }
 
-impl<'a, F: PrimeField> FirstOracles<'a, F> {
+impl<'a, F: PrimeField> SingleEntry<'a, F> {
     /// Iterate over the polynomials output by the prover in the first round.
     /// Intended for use when committing.
     pub fn iter_for_commit(&'a self) -> impl Iterator<Item = LabeledPolynomialWithBasis<'a, F>> {
-        [
-            Some(&self.w_poly).map(Into::into),
-            Some(self.z_a.clone()),
-            Some(self.z_b.clone()),
-            self.mask_poly.as_ref().map(Into::into),
-        ]
-        .into_iter()
-        .flatten()
+        [(&self.w_poly).into(), self.z_a.clone(), self.z_b.clone()].into_iter()
     }
 
     /// Iterate over the polynomials output by the prover in the first round.
     /// Intended for use when opening.
     pub fn iter_for_open(&self) -> impl Iterator<Item = &LabeledPolynomial<F>> {
-        [Some(&self.w_poly), Some(&self.z_a_poly), Some(&self.z_b_poly), self.mask_poly.as_ref()].into_iter().flatten()
+        [(&self.w_poly), &self.z_a_poly, &self.z_b_poly].into_iter()
+    }
+
+    pub fn matches_info(&self, info: &BTreeMap<PolynomialLabel, PolynomialInfo>) -> bool {
+        self.w_poly.info() == &info[self.w_poly.label()]
+            && self.z_a.info() == &info[self.z_a.label()]
+            && self.z_b.info() == &info[self.z_b.label()]
+            && self.z_a_poly.info() == &info[self.z_a_poly.label()]
+            && self.z_b_poly.info() == &info[self.z_b_poly.label()]
```

Comment:
the same suggestion as https://github.com/AleoHQ/snarkVM/pull/760/files#r857827628 applies here and in the other cases of indexing in this file, if there is any chance that it might fail